### ISSUE TITLE
Weird behavior of magicblast: read from the reference can not be aligned to the same reference genome

### ISSUE THREAD

**Issue by xie186** [ncbi/magicblast#8]
I generated a read from reference genome. But the read could be aligned to the genome by magicblast. Could you please help me on what's going on here? 

```
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz
gunzip human_g1k_v37.fasta.gz

## Generate a new reference genome: human_g1k_v37.fasta.blastdb.c2t 
## Create database files for `human_g1k_v37.fasta.blastdb.c2t `
python ../mbmeth.py makeblastdb -r human_g1k_v37.fasta

## Extract a sequence from the new reference 
samtools faidx human_g1k_v37.fasta.blastdb.c2t f18:42015784-42015883 > test_f18_42015784_100.fa

## Align the extracted read to the new reference
magicblast -db human_g1k_v37.fasta.blastdb.c2t -query test_f18_42015784_100.fa > test_f18_42015784_100.mb2geno.sam
```

Here is the alignment result (header not shown here): 
```
@PG     ID:magicblast   PN:magicblast   CL:magicblast -db /home/xie186/WIC_projects/reference_data/human_genome/hg19/human_g1k_v37.fasta.blastdb.c2t -query test_f18_42015784_100.fa
f18:42015784-42015883   4       *       0       0       *       *       0       0       GTTTTTTGTTTATTAAATATGATGTTAGTTGTGGGTTTGTTATATATGATTTTTATTGTGTTGAGATATATTTTTTTAATATATAATTTGTTGAGTTTTT *
```

**Comment by xie186** [comment]
Please ignore this issue for the time being.

**Comment by xie186** [comment]
Here is the command lines I used: 
```
wget http://ftp.1000genomes.ebi.ac.uk/vol1/ftp/technical/reference/human_g1k_v37.fasta.gz
gunzip human_g1k_v37.fasta.gz

## Simulate the reads
wgsim -S 0 ~/WIC_projects/reference_data/human_genome/hg19/human_g1k_v37.fasta out.read1.fq out.read2.fq

## 
makeblastdb -in human_g1k_v37.fasta  -input_type fasta -dbtype nucl -parse_seqids

## Align the reads using magicblast
magicblast -query out.read1.fq -db human_g1k_v37.fasta -query_mate out.read2.fq -splice F  -num_threads 22 -out wgsim_reads2hg19.sam -infmt fastq
```

There are some unmapped reads. What could be the reasons for this? Thanks.

Here is an example of unmaped reads:

```
9_20589011_20589496_0:0:0_0:0:0_7963    69      *       0       0       *       *       0       0       TGGAAGACAGTGTGGCGATTCCTCAGGGATCTAGAACTAGAAATACCATTTGACCCAGCCATCCCATTAC  *
9_20589011_20589496_0:0:0_0:0:0_7963    133     *       0       0       *       *       0       0       CTAACTCGTCATCTAGCATTAGGTATATCTCCCAATGCTATCCCTCCCCCCTCCCCCCACCCCACAACAG  *

```

**Comment by boratyng** [comment]
Magic-BLAST does not align these reads because all 16-base words in these reads repeat more than 60 times in the genome. This is how magicblast avoids aligning to repeats.

You can turn off this filtering with `-limit_lookup F`. Then your first read `9_20589011_20589496_0:0:0_0:0:0_7963` is aligned 256 times. So this really is a repeat.

**Comment by xie186** [comment]
Got it. Thanks.

**Comment by xie186** [comment]
It seems that with `-limit_lookup T` Magic-BLAST is very slow. 

I'm wondering is it possible to add one option to allow us to output the first 60 hits when there is more than 60. Thanks in advance.

**Comment by boratyng** [comment]
Did you mean `-limit_lookup F`? `-limit_lookup T` is the default. Yes, if you use `F` magicblast will be very slow and your are likely going to run out of memory. This is because you will be aligning to all repeated sequences in the genome and you may get hundreds or thousands of alignments for some reads. This option is intended to be used only in special cases (for example when aligning to transcripts).

We can add the option you mention, but need more information: What is your use case? Do you need to see the 60 alignments or do you just need to know that the read aligns somewhere?

**Comment by xie186** [comment]
Thanks for your prompt reply. 

Yes. I mean `-limit_lookup F`. Currently I want to use 'Magic-BLAST' to help me to align short reads to nt database to check the source (species or kingdom) of contamination. Current with `-limit_lookup T`, lots of the reads are shown as unmapped. Especially for BS-seq read alignment in which Cs (Gs) will be converted to Ts (As) the reference will have reduce sequence complexity because of the C->T and G->A conversion. We're writing a tool called `magicblast-meth` (https://github.com/xie186/magicblast-meth) to use `magicblast` to align BS-seq reads against nt database. Then based on the hits we can determine the source of contamination (use `blastcmd` to retrieve the species or kingdom information).  So we'd like `Magic-BLAST` to output some hits even if  the 16-base words in the reads repeat more than 60 times in the reference. 

Thanks.

**Comment by boratyng** [comment]
One more question. It seems to me that the 60 alignments for reads that align more than 60 times may give you wrong results, because you do not know what else these reads align to. Are you correcting for this somehow, for example using uniquely aligned reads?

**Comment by xie186** [comment]
I read the paper. My understanding is that first MagicBLAST only use `words` (16 nt) to search against the reference. `Those words that appear in the reference database more than
a user-specified number of times (by default 60) are not indexed in the lookup table, so that they
never form a seed alignment. ` Basically it doesn't mean there are 60 hits. 

In `magicblast-meth`, we used a similar strategy as the figure shown below to align BS-seq reads:

![image](https://user-images.githubusercontent.com/20909751/63204024-b0ad0300-c061-11e9-9ee6-93ec8e0e0e4e.png)

As you can see there, we convert Cs to Ts in both the reads (G->A conversion for read2) and reference. This means that if we still use 16-nt words, there will be higher possibility that words that appear in the reference database more than 60 times compared to normal read alignment.  So a better idea to solve this problem is maybe you can add one options to change the size of words. For example, if we can have a 32-nt word, then there will be lower possibility of `more than 60 times`. 

Also I noticed a sentence "Magic-BLAST specifically masks out 16-base words that contain at least 15 A’s or 15 T’s, effectively avoiding seeding on polyA tails." In the converted reads, probability with 15 T's (read1 has C->T conversion) or 15 A's (read2 has G->A conversion) will be also higher. So it'd be great that we can have an option to change this parameter. 

In summary, I hope two options could be added to set the size of the word and set the size of polyA(T). 

Please let me know if you have any other information.

**Comment by boratyng** [comment]
Thanks for the explanations. Your understanding about number of word matches is correct. We will try to implement the options that you suggested.

**Comment by xie186** [comment]
Thanks.
